rasa/core/policies/ensemble.py

```diff
@@ -91,590 +220,318 @@ def _check_for_important_policies(self) -> None:
-        self,
-        tracker: DialogueStateTracker,
-        domain: Domain,
-        interpreter: NaturalLanguageInterpreter,
-    ) -> PolicyPrediction:
-        """Finds the best policy prediction.
+        Note that this comparison is *not* symmetric if the priorities are allowed to
+        coincide (i.e. if we cannot distinguish two predictions using 1.-4., then
+        the first prediction is considered to be "better").
 
         Args:
-            tracker: the :class:`rasa.core.trackers.DialogueStateTracker`
-            domain: the :class:`rasa.shared.core.domain.Domain`
-            interpreter: Interpreter which may be used by the policies to create
-                additional features.
-
+          predictions: all predictions made by an ensemble of policies
         Returns:
-            The winning policy prediction.
+          index of the best prediction
         """
-        # find rejected action before running the policies
-        # because some of them might add events
-        rejected_action_name = None
-        last_action_event = next(
-            (
-                event
-                for event in reversed(tracker.events)
-                if isinstance(event, (ActionExecutionRejected, ActionExecuted))
-            ),
-            None,
+        contains_no_form_predictions = all(
+            not prediction.is_no_user_prediction for prediction in predictions
         )
 
-        if len(tracker.events) > 0 and isinstance(
-            last_action_event, ActionExecutionRejected
-        ):
-            rejected_action_name = last_action_event.action_name
-
-        predictions = {
-            f"policy_{i}_{type(p).__name__}": p.predict_action_probabilities(
-                tracker, domain, interpreter
+        def scores(prediction: PolicyPrediction2) -> Tuple[bool, bool, float, int]:
+            """Extracts scores ordered by importance where larger is better."""
+            return (
+                prediction.is_no_user_prediction,
+                contains_no_form_predictions and prediction.is_end_to_end_prediction,
+                prediction.max_probability,
+                prediction.policy_priority,
             )
-            for i, p in enumerate(self.policies)
-        }
 
-        if rejected_action_name:
-            logger.debug(
-                f"Execution of '{rejected_action_name}' was rejected. "
-                f"Setting its confidence to 0.0 in all predictions."
-            )
-            for prediction in predictions.values():
-                prediction.probabilities[
-                    domain.index_for_action(rejected_action_name)
-                ] = 0.0
-
-        return self._pick_best_policy(predictions)
+        # grab the index of a prediction whose tuple of scores is >= all the
+        # tuples of scores for any other prediction
+        arg_max = max(
+            list(range(len(predictions))), key=lambda idx: scores(predictions[idx])
+        )
+        return arg_max
```

Comment:
Just checking: Is this exactly equivalent to the 2.8.x procedure?

Reply:
Can we add some tests that compare outputs of both old and new policy ensembles?

Reply:
The functionality from this function was basically [pick_best_policy](https://github.com/RasaHQ/rasa/blob/0aa9e6c2357dc133d73e870e0d53966dca88fdfa/rasa/core/policies/_ensemble.py#L501) before. There are no tests for that `pick_best_policy` alone in the [tests](https://github.com/RasaHQ/rasa/blob/main/tests/core/test_ensemble.py).

What we had in the old tests: 
- test that policy with higher priority wins if probablities are equal [here](https://github.com/RasaHQ/rasa/blob/main/tests/core/test_ensemble.py#L148)
- end to end predictions supersede others [here](https://github.com/RasaHQ/rasa/blob/main/tests/core/test_ensemble.py#L329)
- no user prediction supersedes others - even if there is an end to end prediction [here](https://github.com/RasaHQ/rasa/blob/main/tests/core/test_ensemble.py#L354)
- tests that check whether optional/mandatory events have been applied (`applies_*_events_*` )
- tests that check whether the featurization definition even thas been added ([here and following test](https://github.com/RasaHQ/rasa/blob/main/tests/core/test_ensemble.py#L434))

All of these have been replaced by equivalent tests ( and we *must* replace them because those tests worked with Policies whereas this new component is supposed to work with PolicyPredictions only)

What is *not* covered:
- the tests related to UnexpecTEDIntent - which rather seem to test what kind of prediction UnexpecTEDIntent makes in a certain situation (because of this [comment](https://github.com/RasaHQ/rasa/blob/main/tests/core/test_ensemble.py#L562))
- ( similar here](https://github.com/RasaHQ/rasa/blob/main/tests/core/test_ensemble.py#L503)

Reply:
@JEM-Mosig  - which tests are you missing and does the above explanation help?

Reply:
added this as TODO for later here: https://www.notion.so/rasa/Cleanup-ensemble-logic-e3a4b699fc2c4c75a8d77613ae31b526